hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java

```diff
@@ -0,0 +1,270 @@
+
+import java.io.IOException;
+import java.util.zip.Checksum;
+import java.util.zip.Deflater;
+import java.util.zip.GZIPOutputStream;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.compress.Compressor;
+import org.apache.hadoop.io.compress.DoNotPool;
+import org.apache.hadoop.util.DataChecksum;
+
+/**
+ * A {@link Compressor} based on the popular gzip compressed file format.
+ * http://www.gzip.org/
+ */
+@DoNotPool
+public class BuiltInGzipCompressor implements Compressor {
+
+  /**
+   * Fixed ten-byte gzip header. See {@link GZIPOutputStream}'s source for
+   * details.
+   */
+  private static final byte[] GZIP_HEADER = new byte[]{
+          0x1f, (byte) 0x8b, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+  private byte[] GZIP_TRAILER = new byte[]{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+  private final int GZIP_HEADER_LEN = 10;
+  private final int GZIP_TRAILER_LEN = 8;
+
+  private Deflater deflater;
+
+  private int headerOff = 0;
+  private int trailerOff = 0;
+
+  private byte[] userBuf = null;
+  private int userBufOff = 0;
+  private int userBufLen = 0;
+
+  private int headerBytesWritten = 0;
+  private int trailerBytesWritten = 0;
+
+  private int currentBufLen = 0;
+
+  private final Checksum crc = DataChecksum.newCrc32();
+
+  private BuiltInGzipDecompressor.GzipStateLabel state;
+
+  public BuiltInGzipCompressor(Configuration conf) { init(conf); }
+
+  @Override
+  public boolean finished() {
+    // Only if the trailer is also written, it is thought as finished.
+    return deflater.finished() && state == BuiltInGzipDecompressor.GzipStateLabel.FINISHED;
+  }
+
+  @Override
+  public boolean needsInput() {
+    if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      return deflater.needsInput();
+    }
+
+    return (state != BuiltInGzipDecompressor.GzipStateLabel.FINISHED);
+  }
+
+  @Override
+  public int compress(byte[] b, int off, int len) throws IOException {
+    int numAvailBytes = 0;
+
+    // If we are not within uncompressed data yet, output the header.
+    if (state != BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      if (userBufLen <= 0) {
+        return numAvailBytes;
+      }
+
+      int outputHeaderSize = writeHeader(b, off, len);
+      headerBytesWritten += outputHeaderSize;
+
+      // Completes header output.
+      if (headerOff == GZIP_HEADER_LEN) {
+        state = BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM;
+      }
+
+      numAvailBytes += outputHeaderSize;
+
+      if (outputHeaderSize == len) {
+        return numAvailBytes;
+      }
+
+      off += outputHeaderSize;
+      len -= outputHeaderSize;
+    }
+
+    if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      // hand off user data (or what's left of it) to Deflater--but note that
+      // Deflater may not have consumed all of previous bufferload, in which case
+      // userBufLen will be zero
+      if (userBufLen > 0) {
```

Comment:
I'm wondering if it's better to do this in `setInput` rather than here. Maybe:

```java
  public void setInput(byte[] b, int off, int len) {
    if (b == null) {
      throw new NullPointerException();
    }
    if (off < 0 || len < 0 || off > b.length - len) {
      throw new ArrayIndexOutOfBoundsException();
    }

    deflater.reset();
    crc.reset();

    deflater.setInput(b, off, len);
    crc.update(b, off, len);  // CRC-32 is on uncompressed data
    currentBufLen = len;
  }
```

with this we can also get rid of `userBuf`, `userBufOff` and `userBufLen`.

Reply:
We need `userBufLen` to record if we have new input for `deflater`. We set it to 0 once we call `setInput` at `deflater`. It is possibly that the deflater doesn't consume all input and we will call `compress` later with partial input in the user buffer. At the moment, `userBufLen` is 0 and we don't call `setInput` at `deflate` again.

Reply:
Why can't the logic in `compress` like this:
```java
    if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
      // now compress it into b[]
      int deflated = deflater.deflate(b, off, len);

      numAvailBytes += deflated;
      off += deflated;
      len -= deflated;
  ..
```
if the `deflater` hasn't consumed all the inputs, `deflater.needsInput()` will return false and next time it'll call `compress` again here.

Reply:
Oh, let me see. Yea, seems working. The logic I used is from decompressor. For decompressor, it needs to parse the header from user input before the inflater can consume user input. So it cannot directly set the input to inflater. But for compressor case, we don't need it and can directly set the input buffer to the deflater at `setInput`.